File: results/dnd-rest-benchmark/lifecycle-runs/20260731T062307Z_claude_claude-sonnet-5_typescript-node/src/routes/characters.ts
// Stateless character math: ability modifiers, proficiency bonus, and the
// combined derived-stats (HP/AC) calculation. No persistence.
import type { ServerResponse } from "node:http";
import { sendJson } from "../http.js";
import { isPlainObject, isValidInt } from "../validation.js";
import { abilityModifier, proficiencyBonus } from "../domain/rules.js";

export function handleAbilityModifier(res: ServerResponse, body: unknown): void {
  if (!isPlainObject(body) || !isValidInt(body.score, 1, 30)) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  sendJson(res, 200, { score: body.score, modifier: abilityModifier(body.score) });
}

export function handleProficiency(res: ServerResponse, body: unknown): void {
  if (!isPlainObject(body) || !isValidInt(body.level, 1, 20)) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  sendJson(res, 200, { level: body.level, proficiency_bonus: proficiencyBonus(body.level) });
}

interface Abilities {
  str: number;
  dex: number;
  con: number;
  int: number;
  wis: number;
  cha: number;
}

const ABILITY_KEYS: (keyof Abilities)[] = ["str", "dex", "con", "int", "wis", "cha"];

export function handleDerivedStats(res: ServerResponse, body: unknown): void {
  if (!isPlainObject(body) || !isValidInt(body.level, 1, 20)) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (!isPlainObject(body.abilities)) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  const abilitiesInput = body.abilities;
  for (const key of ABILITY_KEYS) {
    if (typeof abilitiesInput[key] !== "number" || !Number.isInteger(abilitiesInput[key])) {
      sendJson(res, 400, { error: "invalid request" });
      return;
    }
  }

  if (!isPlainObject(body.armor) || typeof body.armor.base !== "number" || typeof body.armor.dex_cap !== "number") {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  const abilities = abilitiesInput as unknown as Abilities;
  const modifiers = {} as Abilities;
  for (const key of ABILITY_KEYS) {
    modifiers[key] = abilityModifier(abilities[key]);
  }

  const level = body.level as number;
  const bonus = proficiencyBonus(level);
  const hpMax = level * (6 + modifiers.con);
  const shieldBonus = body.armor.shield === true ? 2 : 0;
  const armorClass = body.armor.base + Math.min(modifiers.dex, body.armor.dex_cap) + shieldBonus;

  sendJson(res, 200, {
    level,
    proficiency_bonus: bonus,
    hp_max: hpMax,
    armor_class: armorClass,
    modifiers,
  });
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T070752Z_claude_claude-sonnet-5_ruby-stdlib/lib/handlers/characters.rb
# frozen_string_literal: true

require_relative '../errors'
require_relative '../game_rules'

module Handlers
  # Character-sheet math: ability modifiers, proficiency bonus, and the
  # derived stats (HP, AC) computed from them.
  module Characters
    ABILITY_KEYS = %w[str dex con int wis cha].freeze

    module_function

    def ability_modifier(body)
      score = body['score']
      [200, { score: score, modifier: GameRules.ability_modifier(score) }]
    end

    def proficiency(body)
      level = body['level']
      [200, { level: level, proficiency_bonus: GameRules.proficiency_bonus(level) }]
    end

    def derived_stats(body)
      level = body['level']
      abilities = body['abilities']
      armor = body['armor']

      raise HttpError.new(400, 'level must be an integer') unless level.is_a?(Integer)
      raise HttpError.new(400, 'abilities must be an object') unless abilities.is_a?(Hash)
      raise HttpError.new(400, 'armor must be an object') unless armor.is_a?(Hash)

      bonus = GameRules.proficiency_bonus(level)

      ABILITY_KEYS.each do |key|
        raise HttpError.new(400, "abilities.#{key} must be an integer") unless abilities[key].is_a?(Integer)
      end
      modifiers = ABILITY_KEYS.to_h { |key| [key, GameRules.ability_modifier(abilities[key])] }

      armor_base = armor['base']
      dex_cap = armor['dex_cap']
      shield = armor['shield']

      raise HttpError.new(400, 'armor.base must be an integer') unless armor_base.is_a?(Integer)
      raise HttpError.new(400, 'armor.dex_cap must be an integer') unless dex_cap.is_a?(Integer)
      raise HttpError.new(400, 'armor.shield must be a boolean') unless [true, false].include?(shield)

      hp_max = level * (6 + modifiers['con'])
      shield_bonus = shield ? 2 : 0
      armor_class = armor_base + [modifiers['dex'], dex_cap].min + shield_bonus

      [200, {
        level: level,
        proficiency_bonus: bonus,
        hp_max: hp_max,
        armor_class: armor_class,
        modifiers: {
          str: modifiers['str'],
          dex: modifiers['dex'],
          con: modifiers['con'],
          int: modifiers['int'],
          wis: modifiers['wis'],
          cha: modifiers['cha']
        }
      }]
    end
  end
end
